File: NoTomorrow/Features/Dashboard/NextSessionCard.swift
import SwiftUI

/// The one card on the dashboard (its `FocalCard`): next session time, relative day + countdown, bro row, action row.
struct NextSessionCard: View {
    var session: DashboardSession?
    var routineName: String?
    var isPaired: Bool
    var partnerName: String?
    var myState: DayState
    var myTime: Date?
    var partnerState: DayState
    var partnerTime: Date?
    var hasActiveWorkout: Bool
    /// A finished workout with a completed set started today.
    var trainedToday: Bool = false
    var isConfirming: Bool
    var onConfirm: () -> Void
    var onStart: () -> Void
    var onCantMakeIt: () -> Void

    private var sessionIsToday: Bool { session?.isToday ?? false }
    private var isIn: Bool { sessionIsToday && (myState == .confirmed || myState == .attended) }
    private var isOut: Bool { sessionIsToday && myState.isMissedOrCancelled }
    private var partnerIsIn: Bool { partnerState == .confirmed || partnerState == .attended }

    var body: some View {
        FocalCard {
            VStack(alignment: .leading, spacing: 14) {
                titleRow
                heroRow
                Hairline()
                statusRow
                buttonRow
            }
        }
    }

    // MARK: Rows

    private var titleRow: some View {
        HStack {
            Text(eyebrowKey)
                .eyebrow(NT.Colors.ember)
            Spacer()
            if let routineName {
                Text(routineName).font(NT.Fonts.caption).foregroundStyle(NT.Colors.ink2)
            }
        }
    }

    @ViewBuilder
    private var heroRow: some View {
        if let session {
            HStack(alignment: .lastTextBaseline, spacing: 12) {
                Text(Fmt.time(minuteOfDay: session.minuteOfDay))
                    .font(NT.Fonts.display(64))
                    .foregroundStyle(NT.Colors.ink)
                    .tabular()
                    .lineLimit(1)
                    .minimumScaleFactor(0.6)
                VStack(alignment: .leading, spacing: 2) {
                    Text(Fmt.relativeDay(session.date))
                        .font(NT.Fonts.headline)
                        .foregroundStyle(NT.Colors.ink)
                    TimelineView(.periodic(from: .now, by: 60)) { context in
                        Text(countdown(to: session.date, now: context.date))
                            .font(NT.Fonts.subheadline)
                            .foregroundStyle(NT.Colors.ink2)
                            .tabular()
                    }
                }
                .padding(.bottom, 4)
            }
        } else {
            Text("dashboard.noSchedule")
                .font(NT.Fonts.headline)
                .foregroundStyle(NT.Colors.ink)
                .padding(.vertical, 8)
        }
    }

    @ViewBuilder
    private var statusRow: some View {
        if !sessionIsToday {
            HStack(spacing: 10) {
                Image(systemName: "moon.zzz")
                    .font(.system(size: 15, weight: .semibold))
                    .foregroundStyle(NT.Colors.ink2)
                Text("dashboard.restDay")
                    .font(NT.Fonts.subheadlineBold)
                    .foregroundStyle(NT.Colors.ink2)
            }
        } else if !isPaired {
            HStack(spacing: 10) {
                avatarStack
                Text(soloKey)
                    .font(NT.Fonts.subheadlineBold)
                    .foregroundStyle(isIn || isOut ? NT.Colors.ink : NT.Colors.ink2)
                    .lineLimit(2)
                Spacer(minLength: 8)
                if isIn, let myTime { timeStamp(myTime, color: NT.Colors.good) }
            }
        } else {
            HStack(spacing: 10) {
                avatarStack
                Text(broLine)
                    .font(NT.Fonts.subheadlineBold)
                    .foregroundStyle(NT.Colors.ink)
                    .lineLimit(2)
                Spacer(minLength: 8)
                broTimes
            }
        }
    }

    private var buttonRow: some View {
        HStack(spacing: 10) {
            if hasActiveWorkout {
                PrimaryButton(title: "dashboard.resumeWorkout", height: NT.Size.cardButton, action: onStart)
            } else if !sessionIsToday {
                SecondaryButton(title: "dashboard.startWorkout", height: NT.Size.cardButton, action: onStart)
            } else if isIn {
                PrimaryButton(title: "dashboard.startWorkout", height: NT.Size.cardButton, action: onStart)
            } else {
                PrimaryButton(title: "dashboard.imIn", height: NT.Size.cardButton, isEnabled: !isConfirming, action: onConfirm)
            }
            if Self.offersCantMakeIt(sessionIsToday: sessionIsToday, myState: myState, trainedToday: trainedToday) {
                SecondaryButton(title: "dashboard.cantMakeIt", height: NT.Size.cardButton, action: onCantMakeIt)
            }
        }
    }

    /// "Can't make it" is for a session still ahead today: not once I am out, and not once I trained (it would
    /// cancel a day that is already attended).
    static func offersCantMakeIt(sessionIsToday: Bool, myState: DayState, trainedToday: Bool) -> Bool {
        sessionIsToday && !myState.isMissedOrCancelled && myState != .attended && !trainedToday
    }

    // MARK: Keys

    private var eyebrowKey: LocalizedStringKey {
        session == nil || sessionIsToday ? "dashboard.nextSession" : "dashboard.restDay.next"
    }

    private var soloKey: LocalizedStringKey {
        if isIn { return "dashboard.youreIn" }
        if isOut { return "dashboard.youreOut" }
        return "dashboard.noBro"
    }

    // MARK: Bro row pieces

    private var broLine: String {
        let name = partnerName ?? ""
        switch (isIn, partnerIsIn) {
        case (true, true): return String(localized: "dashboard.bothIn")
        case (false, true): return String(format: String(localized: "dashboard.broIsIn"), name)
        case (true, false):
            return partnerState.isMissedOrCancelled
                ? String(format: String(localized: "dashboard.broOut"), name)
                : String(localized: "dashboard.youreIn")
        case (false, false):
            if isOut { return String(localized: "dashboard.youreOut") }
            return partnerState.isMissedOrCancelled
                ? String(format: String(localized: "dashboard.broOut"), name)
                : String(format: String(localized: "dashboard.broNotYet"), name)
        }
    }

    @ViewBuilder
    private var broTimes: some View {
        switch (isIn, partnerIsIn) {
        case (true, true):
            HStack(spacing: 6) {
                Circle().fill(NT.Colors.good).frame(width: 8, height: 8)
                Text([myTime, partnerTime].compactMap { $0 }.map { Fmt.time($0) }.joined(separator: " · "))
                    .font(NT.Fonts.footnote).foregroundStyle(NT.Colors.ink2).tabular()
            }
        case (false, true):
            if let partnerTime { timeStamp(partnerTime, color: NT.Colors.good) }
        case (true, false):
            if let myTime { timeStamp(myTime, color: NT.Colors.good) }
        case (false, false):
            EmptyView()
        }
    }

    private func timeStamp(_ date: Date, color: Color) -> some View {
        HStack(spacing: 6) {
            Circle().fill(color).frame(width: 8, height: 8)
            Text(Fmt.time(date)).font(NT.Fonts.footnote).foregroundStyle(NT.Colors.ink2).tabular()
        }
    }

    private var avatarStack: some View {
        HStack(spacing: -8) {
            participantAvatar(initial: "", background: NT.Colors.surface2, isIn: isIn, isOut: isOut, dimmed: !isIn)
            if isPaired {
                participantAvatar(initial: partnerName ?? "", background: NT.Colors.surface3,
                                  isIn: partnerIsIn, isOut: partnerState.isMissedOrCancelled, dimmed: !partnerIsIn)
            }
        }
    }

    private func participantAvatar(initial: String, background: Color, isIn: Bool, isOut: Bool, dimmed: Bool) -> some View {
        Avatar(initial: initial.isEmpty ? " " : initial, size: 28, background: background, dimmed: dimmed)
            .overlay(Circle().strokeBorder(isIn ? NT.Colors.good : (isOut ? NT.Colors.bad : NT.Colors.surface), lineWidth: 2))
    }

    private func countdown(to date: Date, now: Date) -> String {
        date <= now ? String(localized: "dashboard.now") : Fmt.countdown(to: date)
    }
}

File: android/app/src/main/java/app/notomorrow/feature/dashboard/NextSessionCard.kt
package app.notomorrow.feature.dashboard

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.text.BasicText
import androidx.compose.foundation.text.TextAutoSize
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.layout.LastBaseline
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import app.notomorrow.designsystem.Eyebrow
import app.notomorrow.designsystem.FocalCard
import app.notomorrow.designsystem.Hairline
import app.notomorrow.designsystem.NT
import app.notomorrow.designsystem.NtIcon
import app.notomorrow.designsystem.NtIcons
import app.notomorrow.designsystem.NtText
import app.notomorrow.designsystem.PrimaryButton
import app.notomorrow.designsystem.SecondaryButton
import app.notomorrow.designsystem.TabularText
import app.notomorrow.designsystem.rememberSecondTicker
import app.notomorrow.designsystem.sfIconSize
import app.notomorrow.designsystem.tabular
import app.notomorrow.service.DayState
import app.notomorrow.util.Fmt
import app.notomorrow.util.S
import app.notomorrow.util.rememberNtStrings
import java.time.Instant

/**
 * The one card on the dashboard (its [FocalCard]): next session time, relative day + countdown,
 * bro row, action row — the port of `NextSessionCard` (`Features/Dashboard/NextSessionCard.swift`).
 */
@Composable
fun NextSessionCard(
    state: DashboardUiState,
    onConfirm: () -> Unit,
    onStart: () -> Unit,
    onCantMakeIt: () -> Unit,
    modifier: Modifier = Modifier,
) {
    val session = state.session
    val sessionIsToday = session?.isToday == true
    val isIn = sessionIsToday &&
        (state.myState is DayState.Confirmed || state.myState is DayState.Attended)
    val isOut = sessionIsToday && state.myState.isMissedOrCancelled
    val partnerIsIn = state.partnerState is DayState.Confirmed || state.partnerState is DayState.Attended

    FocalCard(modifier) {
        Column(
            Modifier.fillMaxWidth(),
            verticalArrangement = Arrangement.spacedBy(14.dp),
        ) {
            TitleRow(
                eyebrow = stringResource(
                    when {
                        state.sessionDone -> S.dashboard_todaysSession
                        session == null || sessionIsToday -> S.dashboard_nextSession
                        else -> S.dashboard_restDay_next
                    }
                ),
                routineName = state.routineName,
            )
            HeroRow(session, done = state.sessionDone)
            Hairline()
            StatusRow(
                state = state,
                sessionIsToday = sessionIsToday,
                isIn = isIn,
                isOut = isOut,
                partnerIsIn = partnerIsIn,
            )
            ButtonRow(
                state = state,
                sessionIsToday = sessionIsToday,
                isIn = isIn,
                onConfirm = onConfirm,
                onStart = onStart,
                onCantMakeIt = onCantMakeIt,
            )
        }
    }
}

// ─────────────────────────────────────────────────────────────────────────────
// Rows
// ─────────────────────────────────────────────────────────────────────────────

@Composable
private fun TitleRow(eyebrow: String, routineName: String?) {
    Row(Modifier.fillMaxWidth(), verticalAlignment = Alignment.CenterVertically) {
        Eyebrow(eyebrow, color = NT.Colors.ember)
        // Swift's `HStack { … Spacer() … }` uses the default 8 pt spacing on BOTH sides of
        // the Spacer, so the routine name never comes closer than 16 pt to the eyebrow.
        // The gap has to live on the (non-weighted) name, which Row measures before the
        // weighted spacer — a `widthIn` on the spacer itself would be coerced away to 0.
        Spacer(Modifier.weight(1f))
        if (routineName != null) {
            NtText(
                text = routineName,
                modifier = Modifier.padding(start = TITLE_MIN_GAP),
                style = NT.Fonts.caption,
                color = NT.Colors.ink2,
                maxLines = 1,
            )
        }
    }
}

/** [done]: today's session is trained, so "Done" stands where the countdown would. */
@Composable
private fun HeroRow(session: DashboardSession?, done: Boolean) {
    if (session == null) {
        NtText(
            text = stringResource(S.dashboard_noSchedule),
            modifier = Modifier.padding(vertical = 8.dp),
            style = NT.Fonts.headline,
            color = NT.Colors.ink,
        )
        return
    }
    val strings = rememberNtStrings()
    val nowMillis by rememberSecondTicker(COUNTDOWN_TICK_MILLIS)
    val now = Instant.ofEpochMilli(nowMillis)

    Row(
        Modifier.fillMaxWidth(),
        horizontalArrangement = Arrangement.spacedBy(12.dp),
    ) {
        // The clock is the weighted child so Row measures it LAST, against whatever the
        // day + countdown column left over — SwiftUI's fixed-size VStack keeps its
        // intrinsic width and `minimumScaleFactor(0.6)` shrinks the hero, not the column.
        // (In a 12-hour locale "10:30 PM" at 64 sp would otherwise eat the whole row.)
        HeroTime(
            Fmt.time(session.minuteOfDay),
            Modifier.weight(1f, fill = false).alignBy(LastBaseline),
        )
        Column(
            modifier = Modifier.alignBy(LastBaseline).padding(bottom = 4.dp),
            verticalArrangement = Arrangement.spacedBy(2.dp),
        ) {
            NtText(
                text = Fmt.relativeDay(session.date, strings),
                style = NT.Fonts.headline,
                color = NT.Colors.ink,
            )
            TabularText(
                text = when {
                    done -> stringResource(S.dashboard_sessionDone)
                    !session.at.isAfter(now) -> stringResource(S.dashboard_now)
                    else -> Fmt.countdown(session.at, strings, now)
                },
                style = NT.Fonts.subheadline,
                color = NT.Colors.ink2,
            )
        }
    }
}

/** `display(64)` with `minimumScaleFactor(0.6)` — one line, shrinking to 38.4 sp. */
@Composable
private fun HeroTime(text: String, modifier: Modifier = Modifier) {
    BasicText(
        text = text,
        modifier = modifier,
        style = NT.Fonts.display(HERO_SIZE).tabular().copy(color = NT.Colors.ink),
        maxLines = 1,
        softWrap = false,
        overflow = TextOverflow.Clip,
        autoSize = TextAutoSize.StepBased(
            minFontSize = (HERO_SIZE * HERO_MIN_SCALE).sp,
            maxFontSize = HERO_SIZE.sp,
            stepSize = 0.5.sp,
        ),
    )
}

@Composable
private fun StatusRow(
    state: DashboardUiState,
    sessionIsToday: Boolean,
    isIn: Boolean,
    isOut: Boolean,
    partnerIsIn: Boolean,
) {
    when {
        !sessionIsToday -> Row(
            Modifier.fillMaxWidth(),
            horizontalArrangement = Arrangement.spacedBy(10.dp),
            verticalAlignment = Alignment.CenterVertically,
        ) {
            NtIcon(NtIcons.MoonZzz, size = sfIconSize(15f), tint = NT.Colors.ink2)
            NtText(
                text = stringResource(S.dashboard_restDay),
                style = NT.Fonts.subheadlineBold,
                color = NT.Colors.ink2,
            )
        }

        !state.isPaired -> Row(
            Modifier.fillMaxWidth(),
            horizontalArrangement = Arrangement.spacedBy(10.dp),
            verticalAlignment = Alignment.CenterVertically,
        ) {
            AvatarStack(state, isIn, isOut, partnerIsIn)
            NtText(
                text = stringResource(
                    when {
                        isIn -> S.dashboard_youreIn
                        isOut -> S.dashboard_youreOut
                        else -> S.dashboard_noBro
                    }
                ),
                modifier = Modifier.weight(1f).padding(end = STATUS_MIN_GAP),
                style = NT.Fonts.subheadlineBold,
                color = if (isIn || isOut) NT.Colors.ink else NT.Colors.ink2,
                maxLines = 2,
            )
            if (isIn && state.myTime != null) TimeStamp(state.myTime, NT.Colors.good)
        }

        else -> Row(
            Modifier.fillMaxWidth(),
            horizontalArrangement = Arrangement.spacedBy(10.dp),
            verticalAlignment = Alignment.CenterVertically,
        ) {
            AvatarStack(state, isIn, isOut, partnerIsIn)
            NtText(
                text = broLine(state, isIn, isOut, partnerIsIn),
                modifier = Modifier.weight(1f).padding(end = STATUS_MIN_GAP),
                style = NT.Fonts.subheadlineBold,
                color = NT.Colors.ink,
                maxLines = 2,
            )
            BroTimes(state, isIn, partnerIsIn)
        }
    }
}

@Composable
private fun ButtonRow(
    state: DashboardUiState,
    sessionIsToday: Boolean,
    isIn: Boolean,
    onConfirm: () -> Unit,
    onStart: () -> Unit,
    onCantMakeIt: () -> Unit,
) {
    Row(
        Modifier.fillMaxWidth(),
        horizontalArrangement = Arrangement.spacedBy(10.dp),
    ) {
        val button = Modifier.weight(1f)
        when {
            state.hasActiveWorkout -> PrimaryButton(
                title = stringResource(S.dashboard_resumeWorkout),
                modifier = button,
                height = NT.Size.cardButton,
                onClick = onStart,
            )

            !sessionIsToday -> SecondaryButton(
                title = stringResource(S.dashboard_startWorkout),
                modifier = button,
                height = NT.Size.cardButton,
                onClick = onStart,
            )

            isIn -> PrimaryButton(
                title = stringResource(S.dashboard_startWorkout),
                modifier = button,
                height = NT.Size.cardButton,
                onClick = onStart,
            )

            else -> PrimaryButton(
                title = stringResource(S.dashboard_imIn),
                modifier = button,
                height = NT.Size.cardButton,
                enabled = !state.isConfirming,
                onClick = onConfirm,
            )
        }
        if (offersCantMakeIt(sessionIsToday, state.myState, state.trainedToday)) {
            SecondaryButton(
                title = stringResource(S.dashboard_cantMakeIt),
                modifier = button,
                height = NT.Size.cardButton,
                onClick = onCantMakeIt,
            )
        }
    }
}

/**
 * `NextSessionCard.offersCantMakeIt` — "Can't make it" is for a session still ahead today: not once
 * I am out, and not once I trained (it would cancel a day that is already attended).
 */
internal fun offersCantMakeIt(sessionIsToday: Boolean, myState: DayState, trainedToday: Boolean): Boolean =
    sessionIsToday && !myState.isMissedOrCancelled && myState !is DayState.Attended && !trainedToday

/** `TimelineView(.periodic(by: 60))` — the relative day and the countdown. */
private const val COUNTDOWN_TICK_MILLIS = 60_000L
private const val HERO_SIZE = 64
private const val HERO_MIN_SCALE = 0.6f

/** Swift's implicit 8 pt HStack spacing on both sides of `titleRow`'s `Spacer()`. */
private val TITLE_MIN_GAP = 16.dp

/**
 * `statusRow`'s `Spacer(minLength: 8)` between the two 10 pt HStack gaps: 28 pt in all.
 * The status text fills the weighted slot (so the trailing timestamp stays flush right,
 * as SwiftUI's expanding Spacer keeps it) and reserves the remaining 18 pt as padding.
 */
private val STATUS_MIN_GAP = 18.dp
